internal/controllers/machine/drain/drain.go

```diff
@@ -0,0 +1,353 @@
+		pod.Kind = "Pod"
+		pod.APIVersion = "v1"
+		pods = append(pods, PodDelete{
+			Pod:    pod,
+			Status: status,
+		})
+	}
+	list := &PodDeleteList{items: pods}
+	return list
+}
+
+// EvictPods evicts the pods.
+func (d *Helper) EvictPods(ctx context.Context, podDeleteList *PodDeleteList) EvictionResult {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Sort podDeleteList, this is important so we always deterministically evict Pods and build the EvictionResult.
+	// Otherwise the condition could change with every single reconcile even if nothing changes on the Node.
+	sort.Slice(podDeleteList.items, func(i, j int) bool {
+		return fmt.Sprintf("%s/%s", podDeleteList.items[i].Pod.GetNamespace(), podDeleteList.items[i].Pod.GetName()) <
+			fmt.Sprintf("%s/%s", podDeleteList.items[j].Pod.GetNamespace(), podDeleteList.items[j].Pod.GetName())
+	})
+
+	var podsToTriggerEviction []PodDelete
+	var podsWithDeletionTimestamp []PodDelete
+	var podsToBeIgnored []PodDelete
+	for _, pod := range podDeleteList.items {
+		switch {
+		case pod.Status.Delete && pod.Pod.DeletionTimestamp.IsZero():
+			podsToTriggerEviction = append(podsToTriggerEviction, pod)
+		case pod.Status.Delete:
+			podsWithDeletionTimestamp = append(podsWithDeletionTimestamp, pod)
+		default:
+			podsToBeIgnored = append(podsToBeIgnored, pod)
+		}
+	}
+
+	log.V(2).Info("Drain not completed yet, there are still Pods on the Node that have to be drained",
+		"podsToTriggerEviction", stringPodDeleteList(podsToTriggerEviction),
+		"podsWithDeletionTimestamp", stringPodDeleteList(podsWithDeletionTimestamp),
+	)
+
+	// Trigger evictions for at most 10s. We'll continue on the next reconcile if we hit the timeout.
+	globalTimeout := 10 * time.Second
+	ctx, cancel := context.WithTimeout(ctx, globalTimeout)
+	defer cancel()
+
+	res := EvictionResult{
+		PodsFailedEviction: map[string][]*corev1.Pod{},
+	}
+
+	for _, pd := range podsToBeIgnored {
+		log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pd.Pod))
+		if pd.Status.Reason == PodDeleteStatusTypeWarning && pd.Status.Message != "" {
+			log = log.WithValues("reason", pd.Status.Message)
+		}
+
+		log.V(3).Info("Skip evicting Pod because it should be ignored")
+		res.PodsIgnored = append(res.PodsIgnored, pd.Pod)
+	}
+
+	for _, pd := range podsWithDeletionTimestamp {
+		log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pd.Pod))
+
+		log.V(3).Info("Skip triggering Pod eviction because it already has a deletionTimestamp")
+		res.PodsDeletionTimestampSet = append(res.PodsDeletionTimestampSet, pd.Pod)
+	}
+
+evictionLoop:
+	for _, pd := range podsToTriggerEviction {
+		log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pd.Pod))
+		if pd.Status.Reason == PodDeleteStatusTypeWarning && pd.Status.Message != "" {
+			log = log.WithValues("warning", pd.Status.Message)
+		}
+		ctx := ctrl.LoggerInto(ctx, log)
+
+		select {
+		case <-ctx.Done():
+			// Skip eviction if the global timeout is reached.
+			err := fmt.Errorf("global timeout %s reached", globalTimeout)
```

Comment:
Changed to
```go
err := fmt.Errorf("eviction timeout of %s reached, eviction will be retried", evictionTimeout)
```

Reason is that it might not be exactly the next reconcile + it's a bit shorter